Remove create_new_password parameter in CLI.build_vault_ids

JustinWayland opened the pull request:
I saw the todo on setup_vault_secrets about removing unused arguments and decided to tackle it.

I can't say I have completely removed the unused argument in build_vault_ids, though, because I started worrying that some unknown code outside of __init__ could be relying on its presence in the signature.  As a result, I am not going to remove it in this commit, but work needs to be done here to figure out if anything or anybody is using build_vault_ids and relying on `create_new_password` being present in the signature.

##### SUMMARY

<!--- Describe the change below, including rationale and design decisions -->
Removed a line in `CLI.setup_vault_secrets` that sent a value to a now unused argument in `CLI.build_vault_ids.` I chickened out on removing the unused argument in `build_vault_ids`'s signature because I don't know whether any other code, internal to this project or externally, implicitly relies on `create_new_password` being in the signature.  If whoever reviews this wants me to remove the argument, I will be happy to do so.

JustinWayland committed:
Remove unused argument in call to build_vault_ids

I saw the todo on setup_vault_secrets about removing unused arguments and decided to tackle it.

I can't say I have completely removed the unused argument in build_vault_ids, though, because I started worrying that some unknown code outside of __init__ could be relying on its presence in the signature.  As a result, I am not going to remove it in this commit, but work needs to be done here to figure out if anything or anybody is using build_vault_ids and relying on create_new_password being present in the signature.

mkrizek reviewed (changes requested):
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment).

JustinWayland committed:
Remove trailing whitespace

JustinWayland committed:
Add changelog fragment

JustinWayland commented:
Sorry about missing the changelog fragment part of creating a pull request. I'm not quite sure where to put my change though. I put it under `minor_changes` for now, but it doesn't feel like it belongs there. Would it be better to put it in another category?

JustinWayland committed:
Remove redundant https://

s-hertel commented on lib/ansible/cli/__init__.py:
```diff
@@ -192,6 +192,12 @@ def split_vault_id(vault_id):
         ret = tuple(parts)
         return ret
 
+    # TODO: verify that nothing else is passing an argument to
+    # create_new_password, then remove it from this function. If anything is
+    # using it inside the ansible codebase, remove it as an argument. If
+    # external code depends on create_new_password being part of the signature,
+    # then add documentation stating that create_new_password cannot be removed
+    # due to external user code relying on it.
     @staticmethod
     def build_vault_ids(vault_ids, vault_password_files=None,
                         ask_vault_pass=None, create_new_password=None,
```
It looks like the only other place in our codebase calling this with `create_new_password` is ./test/units/cli/test_cli.py.

The [Python API](https://docs.ansible.com/ansible/latest/dev_guide/developing_api.html#python-api) is not public and from a quick look I don't see any public code on Github using this method. We could deprecate it for a couple releases using `display.deprecated(msg, version='2.19')`, but seems pretty safe to delete.

JustinWayland replied:
I have deprecated it and updated the changelog fragment to match.

JustinWayland committed:
Deprecate unused parameter

Deprecated passing in create_new_password to build_vault_ids, due to being essentially useless.

It is only not being removed due to external code potentially relying on it, in which case it will log a deprecation notice.

JustinWayland committed:
Remove obsolete build_vault_ids tests

Removed tests and parts of tests pertaining to passing in create_new_password to build_vault_ids, which is now deprecated.

JustinWayland commented:
I have now deprecated the passing in of `create_new_password` to `build-vault-ids`.  I have also removed tests in the test suite that were pointless due to `create_new_password` no longer being used.

That being said, I am unsure what to make of the failure of Sanity Test 2, due to the unhelpful output of `Bash exited with code '1'.` I would like some help figuring out what this means so I can fix the problem.

JustinWayland committed:
Correct errors pointed out by tests

- deprecated_features is valid, deprecated_feature is not
- multiline strings are a feature I should be using

s-hertel commented on lib/ansible/cli/__init__.py:
```diff
@@ -192,10 +192,15 @@ def split_vault_id(vault_id):
     def build_vault_ids(vault_ids, vault_password_files=None,
                         ask_vault_pass=None, create_new_password=None,
                         auto_prompt=True):
+        if create_new_password is not None:
+            display.deprecated("create_new_password is not used in build_vault_ids anymore and is only being kept to maintain compatibility,"
+                               " expect it to be removed in a future release.", version="2.19")
```
If there's a third party caller passing `create_new_password` as None, they won't get this deprecation warning, but they will get a TypeError when we delete the kwarg. Here's one way to fix that:

```diff
+def _handle_deprecated_kwarg(func):
+    def wrapper(*args, **kwargs):
+        if len(args) >= 4 or 'create_new_password' in kwargs:
+            display.deprecated("msg", version='2.19')
+        return func(*args, **kwargs)
+    return wrapper
+
+
 class CLI(ABC):
     ''' code behind bin/ansible* programs '''
 
@@ -192,6 +200,7 @@ class CLI(ABC):
         ret = tuple(parts)
         return ret
 
+    @_handle_deprecated_kwarg
     @staticmethod
     def build_vault_ids(vault_ids, vault_password_files=None,
                         ask_vault_pass=None, create_new_password=None,
```

> create_new_password is not used in build_vault_ids anymore ...

It was never used, so this should just indicate that it's unused and will be removed. No need to mention backwards compatibility for private APIs.

JustinWayland replied:
Thanks for the code! Hopefully the deprecation notice will now display if someone *is* using `create_new_password` by passing in `None`.

JustinWayland replied:
Thanks for the fix! Added you as a co-author to this commit for providing the code for the fix and suggesting a proper deprecation notice.

JustinWayland committed:
Correctly handle deprecated parameter

The previous code did not actually display the deprecation notice.

Thanks to @s-hertel for the fix.

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland committed:
Correct style errors pointed out by linter

JustinWayland committed:
Correct more errors from the linter

s-hertel commented on changelogs/fragments/82066.yml:
```diff
@@ -0,0 +1,3 @@
+---
+deprecated_features:
+  - passing create_new_password to build_vault_ids - it was doing nothing anyways, and is only being deprecated instead of being outright removed due to possibly being used by external code (https://github.com/ansible/ansible/pull/82066)
```
I think the wording should avoid giving the impression that the Python API is supported for external use.

```suggestion
deprecated_features:
  - passing the unused argument ``create_new_password`` to ``CLI.build_vault_ids`` in an abundance of caution for 3rd parties daring to use the unsupported Python API (https://github.com/ansible/ansible/pull/82066).
```

JustinWayland replied:
Thanks for the feedback! Definitely using this instead.

s-hertel commented on test/units/cli/test_cli.py:
```diff
@@ -75,32 +67,12 @@ def test_no_vault_ids_auto_prompt_ask_vault_pass(self):
-        res = cli.CLI.build_vault_ids([], auto_prompt=True, create_new_password=True)
-        self.assertEqual(res, ['default@prompt_ask_vault_pass'])
-
-    def test_create_new_password_no_vault_id_ask_vault_pass(self):
-        res = cli.CLI.build_vault_ids([], ask_vault_pass=True,
-                                      create_new_password=True)
-        self.assertEqual(res, ['default@prompt_ask_vault_pass'])
```
I don't see a test passing just `[], ask_vault_pass=True`, can you just remove `create_new_password=True`?

JustinWayland replied:
Done!

s-hertel commented on test/units/cli/test_cli.py:
```diff
@@ -75,32 +67,12 @@ def test_no_vault_ids_auto_prompt_ask_vault_pass(self):
-                                      create_new_password=True)
-        self.assertEqual(res, ['default@prompt_ask_vault_pass'])
-
-    def test_create_new_password_with_vault_ids(self):
-        res = cli.CLI.build_vault_ids(['foo@bar'], create_new_password=True)
-        self.assertEqual(res, ['foo@bar'])
```
This test still seems useful.

JustinWayland replied:
The problem is that `test` already does this without `create_new_password`. Still putting it back in for now, but I wonder if this test is even necessary now.

s-hertel replied:
Oh, I missed that, feel free to remove it again. Thanks for pointing out the existing coverage.

s-hertel commented on test/units/cli/test_cli.py:
```diff
@@ -75,32 +67,12 @@ def test_no_vault_ids_auto_prompt_ask_vault_pass(self):
-    def test_create_new_password_no_vault_ids_password_files(self):
-        res = cli.CLI.build_vault_ids([], vault_password_files=['some-password-file'],
-                                      create_new_password=True)
-        self.assertEqual(res, ['default@some-password-file'])
```
This one too.

JustinWayland replied:
Done!

JustinWayland committed:
Revert "Remove obsolete build_vault_ids tests"

This reverts commit 1a1c5b537baecd874115ac6d715b48d89bcf5cda.

JustinWayland committed:
Change tests to account for argument removal

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland committed:
Change to discourage usage of this Python API

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland committed:
Merge branch 'ansible:devel' into remove-unused-arguments

s-hertel commented on lib/ansible/cli/__init__.py:
```diff
@@ -115,6 +115,15 @@ def initialize_locale():
 
+def _handle_deprecated_kwarg(func):
+    "Handle the deprecation of create_new_password in build_vault_ids."
+    def wrapper(*args, **kwargs):
```
To avoid overriding `build_vault_ids`'s name/docstring with `wrapper`, decorate with functools.wraps
```suggestion
    @functools.wraps
    def wrapper(*args, **kwargs):
```

s-hertel reviewed (approved):
I missed 1 thing, but looks good to me otherwise

JustinWayland committed:
Prevent overriding of build_vault_ids's name/docstring

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland committed:
Merge branch 'remove-unused-arguments' of https://github.com/JustinWayland/ansible into remove-unused-arguments

JustinWayland committed:
Fix error in declaration

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland committed:
Remove redundant test

JustinWayland commented:
So the wrapping appears to have caused the CI to fail. Either someone can tell me what the correct way to do this would have been, or we can just rip out `create_new_password` entirely and dispense with trying to be nice to people who are using the unsupported Python API.

JustinWayland committed:
Just outright remove create_new_password

A test to see if the CI will be fixed if create_new_password is just outright removed instead of just deprecated.

JustinWayland commented:
At the time of writing this, the CI hasn't completed yet, but just outright removing the attempt to be nice to users of the unsupported Python API appears to have made the CI happy again, as several tests have passed that failed when I tried to just deprecate the feature.

I need help figuring out how to update the changelog fragment to reflect this.

s-hertel commented:
> So the wrapping appears to have caused CI to fail ...

The syntax wasn't right
```
>>> import functools
>>> def deco_with_context(func):
...     @functools.wraps(func)
...     def wrapper(*args, **kwargs):
...         return func(*args, **kwargs)
...     return wrapper
... 
>>> def foo():
...     pass
... 
>>> f = deco_with_context(foo)
>>> f.__name__
'foo'
```
But I prefer just deleting the argument like you've done in the latest version since it's simpler (and I looked for and didn't find any 3rd party public code on Github that would be affected).

The changelog could be a `minor_changes` fragment like you had before, say something like `- removed the unused argument ``create_new_password`` from ``CLI.build_vault_ids`` (#PR).`

JustinWayland committed:
Updated changelog fragment.

Co-Authored-By: Sloane Hertel <19572925+s-hertel@users.noreply.github.com>

JustinWayland commented:
Updated the changelog fragment as indicated. All I need is for @mkrizek to approve my changes and this can be merged.

mkrizek commented on lib/ansible/cli/__init__.py:
```diff
@@ -8,6 +8,7 @@
 import locale
 import os
 import sys
+import functools
```
This seems to be a leftover from a previous commit.

JustinWayland replied:
Whoops! Fixing that.

JustinWayland committed:
Remove unused import

JustinWayland committed:
Add documentation of unsupported API

This is to trigger the CI in order to handle flaky tests.

JustinWayland commented:
I am finding it difficult to appease the CI. Can I get some help?

JustinWayland committed:
Revert "Add documentation of unsupported API"

This reverts commit 2df58775034721fad2dceae98b6c893d195faff9.

JustinWayland commented:
@s-hertel You technically still haven't reviewed and approved me removing the parameter according to @ansibot.

s-hertel commented:
@JustinWayland Thanks for working on this!
